### PR TITLE
Improve macOS and Windows platform integration

### LINKED ISSUE #4933: Adjust app icon for OSX 11 Big Sur
## Summary
[TIP]:  # ( DO NOT include screenshots of your actual database! )
[NOTE]: # ( Provide a brief overview of what the new feature is all about )

Like demonstrated on the last week WWDC the new OS X Big Sur changes many things . One is the different style of the icons used for Programms 

## Examples
[NOTE]: # ( Show us a picture or mock-up of your proposal )

## Context
[NOTE]: # ( Why does this feature matter to you? What unique circumstances do you have? )

Look at this page . Apple changed the icon style from round to square 

https://www.cultofmac.com/715717/fantastic-fugly-all-new-app-icons-macos-big-sur/

### LINKED ISSUE #5349: Follow system theme when app theme is set to "auto"
## Overview
When changing the appearance option inside macOS settings, only half of KeepassXC will change to the correct mode.

## Steps to Reproduce
1. Start KeepassXC
2. Make sure View -> Theme is set to "Automatic"
3. Switch the appearance setting inside Settings -> General

## Expected Behavior
The interface should transition to the correct appearance.

## Actual Behavior
Only half of the interface transitions to the correct appearance, leaving some elements barely readable.
https://www.youtube.com/watch?v=Wu-TvI5ij5g (In the video I'm using an experimental version somewhere between 2.6 and 2.6.1, although I checked 2.6.1 and the issue still persists)

## Context
I also have a similar issue when I have it set to automatic, where I sometimes have to set it to something else, back to automatic and then restart KeepassXC to fix it.

```
KeePassXC - Version 2.6.1
Revision: 9a35bba

Qt 5.15.0
Diagnosemodus ist deaktiviert.

Betriebssystem: macOS 10.15
CPU-Architektur: x86_64
Kernel: darwin 19.6.0

Aktivierte Erweiterungen:
- Auto-Type
- Browser-Integration
- SSH-Agent
- KeeShare (bestätigtes und unbestätigtes Teilen)
- YubiKey
- TouchID

Kryptographische Bibliotheken:
 libgcrypt 1.8.5
```

[NOTE]: # ( Pick choices based on your environment )
Operating System: macOS

### PR DESCRIPTION
**Pull request by phoerious**
Changes in this PR (macOS and Windows only at the moment):

- Allow switching between themes without restart (except classic)
- Rework icon loading and recolouring logic to react to theme changes
- Automatically react to light/dark theme change
- Remove explicit selection of monochrome tray icon variant (selected
  automatically now)
- Update theme background colours for Big Sur
- Update application icon to match Big Sur HIG

The tray icon doesn't respond perfectly to theme changes yet on Big Sur, since we need different icons for dark and light theme and cannot simply let the OS recolour the icon for us (we do that, too, but only as an additional fallback). At the moment, there is no signal to listen to that would allow this.

This patch adds a few generic methods to OSUtils for detecting and communicating theme changes, which are only stubs for Linux at the moment and need to be implemented in future commits.

Fixes #4933
Fixes #5349

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )

output

New icon:
icon_256x256

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change that fixes an issue)
- ✅ New feature (change that adds functionality)

### CODE REVIEW

**Comment by droidmonkey** [comment]
Can we add the big sur style app icon as well

**Comment by varjolintu** [comment]
Does not work on Catalina. Only the title bar changes color, the applications stays the same. I also there's no option to switch to classic theme, or switch themes manually.

And the tray icon is just pure white if set to "Colorful":
Screenshot 2020-12-19 at 10 31 51

With Mojave there's a compliation error with:
```
src/gui/osutils/macutils/AppKitImpl.mm:78:50: error: unused parameter 'notification'
- (void) interfaceThemeChanged:(NSNotification*) notification
```

After setting `#pragma unused (notification)` inside the function the code compiles, and theme switching works, but the same colorful tray icon bug remains.

Built with Qt 5.15.2.

**Comment by phoerious** [comment]
Ah, the colorful icon is probably changed by the OS. I need to unset the mask flag if it's not the monochrome version.

**Comment by phoerious** [comment]
@varjolintu I fixed the compiler warning, the icon bug, and another bug I discovered while switching themes.
I do not know why it's not working on Catalina (but Mojave it is?!) and I cannot test it. Would you be willing to play around with delays in MacUtils.cpp and see if that does anything?

**Comment by varjolintu** [comment]
@phoerious Sure. I'll give it a try.

**Comment by phoerious** [comment]
Added Windows support in a separate commit.

**Comment by droidmonkey** [comment]
Are you going to attempt Linux integration before we merge this?

**Comment by phoerious** [comment]
No. I will do that in a separate PR later.

**Comment by stonerl** [comment]
It is working for me in BS. The only problem I have is the missing icon.

Screenshot 2020-12-30 at 21 04 44

**Comment by phoerious** [comment]
You have to package it. If you run the application from the build directory, it won't pick up the icon. That's always been the case, but the default app icon was different pre-big sur.

**Comment by stonerl** [comment]
I see. I was following the build instructions in [INSTALL.md](https://github.com/keepassxreboot/keepassxc/blob/develop/INSTALL.md) instead of the wiki. Thanks. It is working now.

File: src/gui/Application.cpp [hunk 1]
```diff
@@ -127,6 +128,12 @@ Application::Application(int& argc, char** argv)
         qWarning()
             << QObject::tr("The lock file could not be created. Single-instance mode disabled.").toUtf8().constData();
     }
+
+    connect(osUtils, &OSUtilsBase::interfaceThemeChanged, [this]() {
```
Comment (droidmonkey) [inline src/gui/Application.cpp:132]: Even though application object lives throughout the application life cycle, I prefer the connect with context.

```suggestion
    connect(osUtils, &OSUtilsBase::interfaceThemeChanged, this, [this]() {
```

**Review by droidmonkey — APPROVED** [review]

File: src/gui/Icons.cpp [hunk 2]
```diff
@@ -47,45 +65,102 @@ QString Icons::trayIconAppearance() const
...
+    double dpr = !kpxcApp->testAttribute(Qt::AA_UseHighDpiPixmaps) ? 1.0 : painter->device()->devicePixelRatioF();
+#else
+    double dpr = !kpxcApp->testAttribute(Qt::AA_UseHighDpiPixmaps) ? 1.0 : painter->device()->devicePixelRatio();
+#endif
+    QSize pixmapSize = rect.size() * dpr;
 
-    if (!icon.isNull() && !overrideColor.isValid()) {
-        return icon;
+    painter->save();
+    painter->drawPixmap(rect, m_baseIcon.pixmap(pixmapSize, mode, state));
```
Comment (smlu) [inline src/gui/Icons.cpp:128]: I get very ugly icons on HDPI screen on Windows if icon is painted via `drawPixmap` method:
slika

Icons are drawn correctly If I change the icon pixelmap to be painted via `QIcon::paint`  method e.g. `m_baseIcon.paint(painter, rect, Qt::AlignCenter, mode, state)`
slika

**Comment by phoerious** [comment]
Interesting.

**Comment by phoerious** [comment]
Actually, looking very closely I can see the same thing, just way less severe than what you have.

**Comment by smlu** [comment]
It seems the incorrect painting of an icon on Windows is caused by the screen scale settings. When I set the scale to 100% the icons are drawn correctly, and incorrectly If I increase the scale. 
Looking at what is returned by `devicePixelRatioF()` method, it seems it always return `1.0`. No matter what the global screen scale factor is. Setting custom ration via `dpr` var also doesn't fix the incorrect drawing of an icon.

Therefore, I think it's best to pain the icon image directly via `QIcon::paint` method. At least on the Windows, it works for every scale setting.

**Comment by phoerious** [comment]
I have a scaling of 225% and I can barely see jagged edges.

**Comment by smlu** [comment]
I have it set to 125% and it looks like this is the worst scale setting to have. Increasing scaling above 125%, smooths a bit the jagged edges but the icon is still not clear.

**Comment by phoerious** [comment]
Can you submit your fix as a PR?

**Comment by smlu** [comment]
Ok yes, I can make PR.

**Pull request merged**